v4l-utils/utils/dvb/dvb-format-convert.1.in
.TH "dvb-format-convert" 1 "Fri Oct 3 2014" "DVBv5 Utils @PACKAGE_VERSION@" "User Commands"
.SH NAME
dvb-format-convert \- DVBv5 tool for file format conversions
.SH SYNOPSIS
.B dvb-format-convert
[\fIOPTION\fR]...  \fIinput-file\fR \fIoutput-file\fR
.SH DESCRIPTION
dvb-format-convert is a tool meant to convert among different file formats.
It is compliant with version 5 of the DVB API, being capable of representing
all properties on any standard supported by the Linux digital TV drivers.
.PP
There are currently 3 different formats supported for input:
.IP "\(bu" 2
\fBdvbv5\fR \- the standard format at libdvbv5, capable of representing all
different TV standards;
.PP
.IP "\(bu" 2
\fBchannel\fR \- the dvb-apps legacy channel format, with supports only
ATSC, DVB-C, DVB-S and DVB-T standards, extended to also support s2-scan
format for DVB-S2 files, and to support DVB-T2.
.IP "\(bu" 2
\fBzap\fR \- the dvb-apps legacy format for tuning, with supports only
ATSC, DVB-C, DVB-S and DVB-T standards.
.PP
There is one extra output format:
.PP
.IP "\(bu" 2
\fBvdr\fR \- the format used by the vdr application. It supports all features
found up on vdr version 2.1.6.
.PP
Unfortunately, the dvb-apps legacy zap format doesn't store the DVB standard,
so, any conversions to and/or from this format requires an extra parameter,
to specify the delivery system.
.PP
.SH OPTIONS
.TP
The following options are valid:
.TP
\fB-I\fR, \fB--input-format\fR=\fIformat\fR
Format of the input file.
Supported input formats: \fBchannel\f, \fBzap\fR and \fBdvbv5\fR.
.TP
\fB-O\fR, \fB--output-format\fR=\fIformat\fR
Format of the output file.
Supported output formats: \fBvdr\fR, \fBchannel\fR, \fBzap\fR and \fBdvbv5\fR.
.TP
\fB-s\fR, \fB--delsys\fR=\fIsystem\fR
Delivery system type.
Needed if input or output format is ZAP.
.TP
\fB-?\fR, \fB--help\fR
Outputs the usage help.
.TP
\fB--usage\fR
Give a short usage message.
.TP
\fB-V\fR, \fB--version\fR
Print program version.
.SH EXAMPLES
.RS
.SS Converting a channel file from legacy dvb-apps format
Reads a file at the dvb\-apps channel file and writes it at dvbv5 format.
.PP
.nf
$ \fBdvb\-format\-convert \-I channel \-O dvbv5 dvbc\-channel\-legacy dvbc\-channel\fR
Reading file dvbc\-channel\-legacy
Writing file dvbc\-channel
.fi
.SS Convert a dvbv5 file with programs on it to the legacy dvb\-apps zap format
Reads a file generated by dvbv5-scan on dvbv5 format and writes a new file
at dvb\-apps zap format.
.PP
.nf
$ \fBdvb\-format\-convert \-I dvbv5 \-O zap \-s dvbc/annex_a dvb_channel.conf legacy.conf\fR
Reading file dvb_channel.conf
Writing file legacy.conf
WARNING: missing video PID while parsing entry 6 of legacy.conf
WARNING: missing video PID while parsing entry 7 of legacy.conf
WARNING: missing video PID while parsing entry 8 of legacy.conf
WARNING: missing audio PID while parsing entry 8 of legacy.conf
.fi
.PP
Please notice that the legacy dvb\-apps format assumes that every channel
will have one (and just one) audio PID and one (and just one) video PID,
while the dvbv5 format supports multiple PID per service type.
.PP
The above warnings happened on pure\-audio channels and on data\-only channels:
.PP
Original dvb_channel.conf file:
.PP
.nf
[the]
        SERVICE_ID = 4
        VIDEO_PID = 42
        AUDIO_PID = 257
        PID_f1 = 768
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[quick]
        SERVICE_ID = 5
        VIDEO_PID = 288
        AUDIO_PID = 289
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[brown]
        SERVICE_ID = 7
        VIDEO_PID = 36
        AUDIO_PID = 305
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[fox]
        SERVICE_ID = 13
        VIDEO_PID = 272
        AUDIO_PID = 273
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[jumps]
        SERVICE_ID = 18
        VIDEO_PID = 368
        AUDIO_PID = 369
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[over]
        SERVICE_ID = 48
        VIDEO_PID = 336
        AUDIO_PID = 337 338 849
        PID_86 = 816
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[the]
        SERVICE_ID = 57
        AUDIO_PID = 353
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[lazy]
        SERVICE_ID = 104
        AUDIO_PID = 321
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[dog]
        SERVICE_ID = 252
        PID_05 = 3047 3046 3045 3044 3042
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A

.fi
.PP
So, for the channels with multiple PIDs, the converted legacy.conf file
will have just the first one.
.PP
For the channels without an audio or video PID, it will store PID=0.
.PP
This is the output file for the above channel list.
.PP
.nf
the:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:42:257:4
quick:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:288:289:5
brown:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:36:305:7
fox:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:272:273:13
jumps:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:368:369:18
over:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:336:337:48
the:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:0:353:57
lazy:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:0:321:104
dog:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:0:0:252
.fi
.SH BUGS
Report bugs to \fBLinux Media Mailing List <linux-media@vger.kernel.org>\fR
.SH COPYRIGHT
Copyright (c) 2011\-2014 by Mauro Carvalho Chehab.
.PP
License GPLv2: GNU GPL version 2 <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


i2c-tools/CHANGES
i2c-tools CHANGES
-----------------

3.1.3 (work in progress)
  eeprog: Increase delay after writes
  decode-dimms: Correctly check for out-of-bounds vendor ID
  README: Mention the current maintainer

3.1.2 (2015-06-17)
  decode-dimms: Fix DDR3 extended temp range refresh rate decoding
  py-smbus: Add support for python 3

3.1.1 "Happy Birthday Pollux" (2014-02-20)
  decode-dimms: Decode module configuration type of DDR2 SDRAM
                Decode bus width extension of DDR3 SDRAM
                Don't choke when no EEPROM is found
                Update manufacturer IDs
                Remove duplicate "ns" in SDR timings
                Fix decoding of SDR SPD revision
                Prevent hang on reserved DDR3 module type
                Decode more DDR3 module types
                Fix DDR3 tRAS decoding
                Fix DDR3 core timings rounding
                Round down PC3 numbers to comply with Jedec
                Encode "degrees" to HTML degree symbol
                Add a manual page
  decode-vaio: Add a manual page
  eeprog: Add a manual page
  eeprom: Add a manual page
  eepromer: Add a manual page
  i2cdetect: Clarify the SMBus commands used for probing by default
  i2c-dev.h: Minimize differences with kernel flavor
  i2c-stub-from-dump: Be more tolerant on input dump format
  py-smbus: Fix module level docs

3.1.0 (2011-12-04)
  decode-dimms: Decode module configuration type of DDR SDRAM
                Decode refresh rate of DDR SDRAM
                Add support for the at24 kernel driver
  i2c-dev.h: Make value arrays const for block write functions
  i2cset: Add support for SMBus and I2C block writes
          Removed obsolete method to specify value mask
          More stringent parameter validation

3.0.3 (2010-12-12)
  Makefile: Let the environment set CC and CFLAGS
            Integrate py-smbus into the build system
  README: Point users to the linux-i2c mailing list
  decode-dimms: Handle CRC of FB-DIMM and DDR3 SDRAM memory modules
                Add support for DDR3 SDRAM
                Fix decoding of SDR SDRAM bytes 12-14
                Add side-by-side formatting option
                Add merged cells formatting option
                Try harder to decode the manufacturing date
                Handle read errors on sysfs
                Decode voltage interface level of DDR SDRAM
  decode-xeon: Delete
  eepromer: Fix array initialization overrun
  i2cdetect: Drop legacy reference to ISA bus
  i2cset: Add support for short writes with PEC
  i2c-stub-from-dump: Use udev settle to speed up initialization
                      Unload i2c-stub automatically if needed
                      Add support for multiple dumps
  tools: Properly handle /dev/i2c not being a directory
         Increase limit on I2C bus number

3.0.2 (2008-11-29)
  i2c-dev.h: Drop I2C_FUNC_SMBUS_*I2C_BLOCK_2 defines
  decode-dimms: Add support for little-endian word hexdumps
                Only export the ceil function from POSIX
  decode-vaio: Remove history
  i2cdetect: Support i2c bus passed by name
             Shorten the usage message
  i2cdump: Support i2c bus passed by name
           Shorten the usage message
           Restrict the chip address to 0x03-0x77
           Split the functionality checking code into a separate function
           Better error message on missing adapter functionality
  i2cget: Support i2c bus passed by name
          Shorten the usage message
          Better error message on missing adapter functionality
  i2cset: Support i2c bus passed by name
          Shorten the usage message
          Restrict the chip address to 0x03-0x77
          Split the code into several functions for clarity
          Add support for short writes (SMBus send byte)
          Better error message on missing adapter functionality
          Set the data value mask with -m
          Make reading back the written value optional
  i2c-stub-from-dump: Add support for partial dumps
                      Report if only garbage is found in dump file
                      Behave properly when i2c-stub is already loaded
                      Stop on i2cset error

3.0.1 (2008-04-04)
  Drop the trailing .pl from all perl script names
  decode-dimms: Fix DDR2 SDRAM module speed decoding
                Update manufacturer IDs
                Don't print anything by default if checksum fails
                Decode all DDR2 SDRAM timing information
                Add support for reading SPD data from hexdumps
                Make command line parsing more robust
  decode-vaio: Private data might not be readable by non-root users
               Print the asset tag
               Fix the timestamp decoding
  i2cdump: Fix I2C block mode error code
           Remove man page reference to hardware monitoring chips
           Let the user specify a register range
  i2cset: Final status messages go to stdout
          Return success even when readback fails or doesn't match
  i2c-stub-from-dump: New helper script to use with i2c-stub

3.0.0 (2007-10-14)
  Initial release


v4l-utils/utils/dvb/dvb-format-convert.c
/*
 * Copyright (c) 2011-2012 - Mauro Carvalho Chehab
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation version 2
 * of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 * Or, point your browser to http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 *
 */

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>
#include <argp.h>

#include <config.h>

#ifdef ENABLE_NLS
# define _(string) gettext(string)
# include "gettext.h"
# include <locale.h>
# include <langinfo.h>
# include <iconv.h>
#else
# define _(string) string
#endif

# define N_(string) string

#include "libdvbv5/dvb-file.h"
#include "libdvbv5/dvb-demux.h"
#include "libdvbv5/dvb-scan.h"

#define PROGRAM_NAME	"dvb-format-convert"

struct arguments {
	char *input_file, *output_file;
	enum dvb_file_formats input_format, output_format;
	int delsys;
};

static const struct argp_option options[] = {
	{"input-format",	'I',	N_("format"),	0, N_("Valid input formats: ZAP, CHANNEL, DVBV5"), 0},
	{"output-format",	'O',	N_("format"),	0, N_("Valid output formats: VDR, ZAP, CHANNEL, DVBV5"), 0},
	{"delsys",		's',	N_("system"),	0, N_("Delivery system type. Needed if input or output format is ZAP"), 0},
	{"help",        '?',	0,		0,	N_("Give this help list"), -1},
	{"usage",	-3,	0,		0,	N_("Give a short usage message")},
	{"version",	'V',	0,		0,	N_("Print program version"), -1},
	{ 0, 0, 0, 0, 0, 0 }
};

const char *argp_program_version = PROGRAM_NAME " version " V4L_UTILS_VERSION;
const char *argp_program_bug_address = "Mauro Carvalho Chehab <m.chehab@samsung.com>";

static error_t parse_opt(int k, char *optarg, struct argp_state *state)
{
	struct arguments *args = state->input;
	switch (k) {
	case 'I':
		args->input_format = dvb_parse_format(optarg);
		break;
	case 'O':
		args->output_format = dvb_parse_format(optarg);
		break;
	case 's':
		args->delsys = dvb_parse_delsys(optarg);
		break;
	case '?':
		argp_state_help(state, state->out_stream,
				ARGP_HELP_SHORT_USAGE | ARGP_HELP_LONG
				| ARGP_HELP_DOC);
		fprintf(state->out_stream, _("\nReport bugs to %s.\n"), argp_program_bug_address);
		exit(0);
	case 'V':
		fprintf (state->out_stream, "%s\n", argp_program_version);
		exit(0);
	case -3:
		argp_state_help(state, state->out_stream, ARGP_HELP_USAGE);
		exit(0);
	default:
		return ARGP_ERR_UNKNOWN;
	};
	return 0;
}

static int convert_file(struct arguments *args)
{
	struct dvb_file *dvb_file = NULL;
	int ret;

	printf(_("Reading file %s\n"), args->input_file);

	dvb_file = dvb_read_file_format(args->input_file, args->delsys,
				    args->input_format);
	if (!dvb_file) {
		fprintf(stderr, _("Error reading file %s\n"), args->input_file);
		return -1;
	}

	printf(_("Writing file %s\n"), args->output_file);
	ret = dvb_write_file_format(args->output_file, dvb_file,
				    args->delsys, args->output_format);
	dvb_file_free(dvb_file);

	return ret;
}

int main(int argc, char **argv)
{
	struct arguments args;
	int idx = -1, missing = 0;
	const struct argp argp = {
		.options = options,
		.parser = parse_opt,
		.doc = N_("scan DVB services using the channel file"),
		.args_doc = N_("<input file> <output file>"),
	};

#ifdef ENABLE_NLS
	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);
#endif

	memset(&args, 0, sizeof(args));
	argp_parse(&argp, argc, argv, ARGP_NO_HELP | ARGP_NO_EXIT, &idx, &args);

	if (idx + 1 < argc) {
		args.input_file = argv[idx];
		args.output_file = argv[idx + 1];
	}

	if (args.input_format == FILE_UNKNOWN) {
		fprintf(stderr, _("ERROR: Please specify a valid input format\n"));
		missing = 1;
	} else if (!args.input_file) {
		fprintf(stderr, _("ERROR: Please specify a valid input file\n"));
		missing = 1;
	} else if (args.output_format == FILE_UNKNOWN) {
		fprintf(stderr, _("ERROR: Please specify a valid output format\n"));
		missing = 1;
	} else if (!args.output_file) {
		fprintf(stderr, _("ERROR: Please specify a valid output file\n"));
		missing = 1;
	} else if (((args.input_format == FILE_ZAP) ||
		   (args.output_format == FILE_ZAP)) &&
		   (args.delsys <= 0 || args.delsys == SYS_ISDBS)) {
		fprintf(stderr, _("ERROR: Please specify a valid delivery system for ZAP format\n"));
		missing = 1;
	}
	if (missing) {
		argp_help(&argp, stderr, ARGP_HELP_STD_HELP, PROGRAM_NAME);
		return -1;
	}

	return convert_file(&args);
}
